prow/cron/cron.go

```diff
@@ -0,0 +1,172 @@
+limitations under the License.
+*/
+
+// Package cron provides a wrapper of robfig/cron, which manages schedule cron jobs for horologium
+package cron
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/robfig/cron"
+
+	"k8s.io/test-infra/prow/config"
+)
+
+// jobStatus is a cache layer for tracking existing cron jobs
+type jobStatus struct {
+	entryID cron.EntryID
+	trigger bool
+	cronStr string
+}
+
+// Cron is a wrapper for cron.Cron
+type Cron struct {
+	cronAgent *cron.Cron
+	jobs      map[string]*jobStatus
+	lock      sync.Mutex
+}
+
+// New makes a new Cron object
+func New() *Cron {
+	return &Cron{
+		cronAgent: cron.New(),
+		jobs:      map[string]*jobStatus{},
+	}
+}
+
+// Start kicks off current cronAgent scheduler
+func (c *Cron) Start() {
+	c.cronAgent.Start()
+}
+
+// Stop pauses current cronAgent scheduler
+func (c *Cron) Stop() {
+	c.cronAgent.Stop()
+}
+
+// QueuedJobs returns a list of jobs that need to be triggered
+// and reset trigger in jobStatus
+func (c *Cron) QueuedJobs() []string {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	res := []string{}
+	for k, v := range c.jobs {
+		if v.trigger {
+			res = append(res, k)
+		}
+		c.jobs[k].trigger = false
+	}
+	return res
+}
+
+// SyncConfig syncs current cronAgent with current prow config
+// which add/delete jobs accordingly.
+func (c *Cron) SyncConfig(cfg *config.Config) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	for _, p := range cfg.Periodics {
+		if err := c.addPeriodic(p); err != nil {
+			return err
+		}
+	}
+
+	exist := map[string]bool{}
+	for _, p := range cfg.AllPeriodics() {
+		exist[p.Name] = true
+	}
+
+	for k := range c.jobs {
+		if _, ok := exist[k]; !ok {
+			defer c.removeJob(k)
+		}
+	}
+
+	return nil
+}
+
+// HasJob returns if a job has been scheduled in cronAgent or not
+func (c *Cron) HasJob(name string) bool {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	_, ok := c.jobs[name]
+	return ok
+}
+
+func (c *Cron) addPeriodic(p config.Periodic) error {
+	if p.Cron == "" {
+		return nil
+	}
+
+	if job, ok := c.jobs[p.Name]; ok {
+		if job.cronStr == p.Cron {
+			return nil
+		}
+		// job updated, remove old entry
+		if err := c.removeJob(p.Name); err != nil {
+			return err
+		}
+	}
+
+	if err := c.addJob(p.Name, p.Cron); err != nil {
+		return err
+	}
+
+	for _, ras := range p.RunAfterSuccess {
+		if err := c.addPeriodic(ras); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// addJob adds a cron entry for a job to cronAgent
+func (c *Cron) addJob(name, cron string) error {
+	id, err := c.cronAgent.AddFunc(cron, func() {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		// We want to ignore second trigger if first trigger is not consumed yet.
+		c.jobs[name].trigger = true
+	})
+
+	if err != nil {
+		return fmt.Errorf("cronAgent fails to add job %s with cron %s: %v", name, cron, err)
+	}
+
+	c.jobs[name] = &jobStatus{
+		entryID: id,
+		// try to kick of a periodic trigger right away
+		trigger: strings.HasPrefix(cron, "@every"),
```

Comment:
@spxtr I added the initial state of trigger base off if the cron string starts with `@every` or not, so that if horologium is down, when it recovers and resyncs the config, the job will be scheduled immediately.

Reply:
For job uses actual cron string, if it has to be that unfortunate, then good luck :-(